Readme.md
# retext-language [![Build Status](https://img.shields.io/travis/wooorm/retext-language.svg?style=flat)](https://travis-ci.org/wooorm/retext-language) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-language.svg?style=flat)](https://coveralls.io/r/wooorm/retext-language?branch=master)

Detect the language of text with **[Retext](https://github.com/wooorm/retext)**.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-language
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-language
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-language
```

[Duo](http://duojs.org/#getting-started):

```javascript
var language = require('wooorm/retext-language');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var language = require('retext-language');

var retext = new Retext()
    .use(visit)
    .use(language);

retext.parse('Alle mennesker er født frie og', function (err, tree) {
    console.log(tree.data.language);
    /* 'nno' // Which is Norwegian Nynorsk. */
});

retext.parse(
    'Tots els éssers humans neixen lliures i iguals. ' + 
    'All human beings are born free and equal.\n' + 
    'Toate ființele umane se nasc libere și egale. ' + 
    '人人生而自由﹐在尊嚴和權利上一律平等。',
    function (err, tree) {
        console.log(tree.data.language);
        /* 'cat' // Catalan */

        tree.visit(tree.PARAGRAPH_NODE, function (node) {
            console.log(node.toString(), node.data.language);
        });
        /*
         * 'Tots els éssers humans neixen lliures i iguals. All human beings are born free and equal.', 'cat' // Catalan
         * 'Toate ființele umane se nasc libere și egale. 人人生而自由﹐在尊嚴和權利上一律平等。', 'ron' // Romanian
         */

        tree.visit(tree.SENTENCE_NODE, function (node) {
            console.log(node.toString(), node.data.language);
        });
        /*
         * 'Tots els éssers humans neixen lliures i iguals.', 'cat' // Catalan
         * 'All human beings are born free and equal.', 'eng' // English
         * 'Toate ființele umane se nasc libere și egale.', 'ron' // Romanian
         * '人人生而自由﹐在尊嚴和權利上一律平等。', 'cmn' // Mandarin Chinese
         */
    }
);
```

## Supported Languages

retext-language supports the 175 languages provided by **[wooorm/franc](https://github.com/wooorm/franc#supported-languages)**. 

**Note!** Results may vary depending on which [parser](https://github.com/wooorm/retext#parsers) you use. It’s probably better for most cases to detect the language (with **[wooorm/franc](https://github.com/wooorm/franc)**) before parsing, and choose a parser according to the results.

## API

None, **retext-language** automatically detects the language of each [`SentenceNode`](https://github.com/wooorm/textom#textomsentencenode-nlcstsentencenode) (using **[wooorm/franc](https://github.com/wooorm/franc)**), and stores the language in `node.data.language`. Additionally, information about all detected languages is stored in `node.data.languages`.

The average of the detected languages on parents ([paragraph](https://github.com/wooorm/textom#textomparagraphnode-nlcstparagraphnode)s and [root](https://github.com/wooorm/textom#textomrootnode-nlcstrootnode)s), through the same `parent.data.language` and `parent.data.languages`.

## Performance

On a MacBook air:

```text
A paragraph (5 sentences, 100 words)
retext w/o retext-language x 215 ops/sec ±1.48% (80 runs sampled)
retext w/ retext-language x 1.52 ops/sec ±1.27% (8 runs sampled)

A section (10 paragraphs, 50 sentences, 1,000 words)
retext w/o retext-language x 20.79 ops/sec ±2.92% (39 runs sampled)
retext w/ retext-language x 0.15 ops/sec ±4.92% (5 runs sampled)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.3.4 / 2015-01-19
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Refactor to use generic `changeinside` in new retext
  * Update retext to stable release

0.3.3 / 2014-12-14
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update franc, eslint, retext

0.3.2 / 2014-11-22
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependencies
 * Update franc, retext

0.3.1 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update franc, eslint, mocha, retext, retext-visit, retext-content

0.3.0 / 2014-10-21
==================

 * Refactor .jscs.json
 * Update test for changes in retext-visit
 * Update module for changes in retext-visit
 * Update retext-visit, franc, retext
 * Update .gitignore, .npmignore, bower ignore
 * Update japanese fixture in test
 * Move spec to test.js

0.2.1 / 2014-10-03
==================

 * Update component.json, bower.json for franc@0.2.0

0.2.0 / 2014-10-03
==================

 * Update spec for a9e7794
 * Update language detection in API
 * Update franc to 0.2.0

0.1.2 / 2014-09-27
==================

 * Refactor Readme.md
 * Refactor API
 * Remove browser test
 * Update .gitignore, .npmignore
 * Fix property order in component.json, package.json, bower.json
 * Remove testling
 * Refactor spec
 * Update dependencies

0.1.1 / 2014-09-08
==================

 * Add install guide for bower
 * Fix casing of NPM > npm
 * Revert wrongly removed code in 58e4177
 * Update retext-content to 0.2.1

0.1.0 / 2014-09-08
==================

 * Add bower.json
 * Update retext to 0.1.1
 * Add components, bower_components to .gitignore
 * Remove `npm update npm` from travis before_install
 * Add npm deployment to travis
 * Update eslint, retext-content


test.js
'use strict';

/*
 * Dependencies.
 */

var Retext,
    language,
    inspect,
    content,
    visit,
    assert;

Retext = require('retext');
language = require('./');
inspect = require('retext-inspect');
content = require('retext-content');
visit = require('retext-visit');
assert = require('assert');

/*
 * Retext.
 */

var retext;

retext = new Retext()
    .use(inspect)
    .use(content)
    .use(visit)
    .use(language);

/*
 * Fixtures.
 */

var root,
    languages,
    paragraphLanguages,
    otherSentences,
    otherLanguages,
    otherParagraphLanguages;

root =
    'Alle menslike wesens word vry. ' +
    'O Brasil caiu 26 posições em.\n' +
    'All human beings are born. ' +
    'Tous les êtres humains naissent.';

languages = ['afr', 'por', 'eng', 'fra'];

paragraphLanguages = [
    languages.slice(0, 2), languages.slice(2)
];

otherSentences = [
    'Bütün insanlar ləyaqət və hüquqlarına görə.',
    'Tots els éssers humans neixen lliures i iguals',
    'چونکہ ہر انسان کی ذاتی عز',
    'ふぁぼとは、その人のツイートをお気に入り登録することです。RT記号の横の星を押せば完了です。'
];

otherLanguages = ['azj', 'cat', 'urd', 'jpn'];

otherParagraphLanguages = [
    otherLanguages.slice(0, 2), otherLanguages.slice(2)
];

/*
 * Tests.
 */

describe('language()', function () {
    var tree;

    before(function (done) {
        retext.parse(root, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof language === 'function');
    });

    it('should process `Parent`', function () {
        var index;

        index = -1;

        tree.visit(tree.SENTENCE_NODE, function (node) {
            index++;

            assert(node.data.language === languages[index]);
        });

        index = -1;

        tree.visit(tree.PARAGRAPH_NODE, function (node) {
            index++;

            assert(paragraphLanguages[index].indexOf(
                node.data.language
            ) !== -1);
        });

        assert(languages.indexOf(tree.data.language) !== -1);
    });

    it('should set `language` to `null` when node has no value', function () {
        tree.visit(tree.SENTENCE_NODE, function (node) {
            node.removeContent();

            assert(node.data.language === null);
        });

        tree.visit(tree.PARAGRAPH_NODE, function (node) {
            assert(node.data.language === null);
        });

        assert(tree.data.language === null);
    });

    it('should re-process nodes when their value changes', function () {
        var index;

        index = -1;

        tree.visit(tree.SENTENCE_NODE, function (node) {
            index++;

            node.replaceContent(otherSentences[index]);

            assert(
                node.data.language === otherLanguages[index]
            );
        });

        index = -1;

        tree.visit(tree.PARAGRAPH_NODE, function (node) {
            index++;

            assert(otherParagraphLanguages[index].indexOf(
                node.data.language
            ) !== -1);
        });

        assert(otherLanguages.indexOf(tree.data.language) !== -1);
    });
});


spec/retext-language.spec.js
'use strict';

/**
 * Dependencies.
 */

var language,
    content,
    visit,
    Retext,
    assert;

language = require('..');
content = require('retext-content');
Retext = require('retext');
visit = require('retext-visit');
assert = require('assert');

/**
 * Retext.
 */

var retext;

retext = new Retext()
    .use(content)
    .use(visit)
    .use(language);

/**
 * Fixtures.
 */

var document,
    languages,
    paragraphLanguages,
    otherSentences,
    otherLanguages,
    otherParagraphLanguages;

document =
    'Alle menslike wesens word vry. ' +
    'O Brasil caiu 26 posições em.\n' +
    'All human beings are born. ' +
    'Tous les êtres humains naissent.';

languages = ['afr', 'por', 'eng', 'fra'];

paragraphLanguages = [
    languages.slice(0, 2), languages.slice(2)
];

otherSentences = [
    'Bütün insanlar ləyaqət və hüquqlarına görə.',
    'Tots els éssers humans neixen lliures i iguals',
    'چونکہ ہر انسان کی ذاتی عز',
    '人類社会のすべての構成員の固有の尊厳と平等で譲るこ'
];

otherLanguages = ['azj', 'cat', 'urd', 'jpn'];

otherParagraphLanguages = [
    otherLanguages.slice(0, 2), otherLanguages.slice(2)
];

/**
 * Tests.
 */

describe('language()', function () {
    var tree;

    before(function (done) {
        retext.parse(document, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof language === 'function');
    });

    it('should process `Parent`', function () {
        var index;

        index = -1;

        tree.visitType(tree.SENTENCE_NODE, function (node) {
            index++;

            assert(node.data.language === languages[index]);
        });

        index = -1;

        tree.visitType(tree.PARAGRAPH_NODE, function (node) {
            index++;

            assert(paragraphLanguages[index].indexOf(
                node.data.language
            ) !== -1);
        });

        assert(languages.indexOf(tree.data.language) !== -1);
    });

    it('should set `language` to `null` when node has no value', function () {
        tree.visitType(tree.SENTENCE_NODE, function (node) {
            node.removeContent();

            assert(node.data.language === null);
        });

        tree.visitType(tree.PARAGRAPH_NODE, function (node) {
            assert(node.data.language === null);
        });

        assert(tree.data.language === null);
    });

    it('should re-process nodes when their value changes', function () {
        var index;

        index = -1;

        tree.visitType(tree.SENTENCE_NODE, function (node) {
            index++;

            node.replaceContent(otherSentences[index]);

            assert(
                node.data.language === otherLanguages[index]
            );
        });

        index = -1;

        tree.visitType(tree.PARAGRAPH_NODE, function (node) {
            index++;

            assert(otherParagraphLanguages[index].indexOf(
                node.data.language
            ) !== -1);
        });

        assert(otherLanguages.indexOf(tree.data.language) !== -1);
    });
});


index.js
/**
 * Dependencies.
 */

var Retext = require('wooorm/retext@0.5.0');
var language = require('wooorm/retext-language@0.3.4');
var debounce = require('component/debounce@1.0.0');

/**
 * Retext.
 */

var retext = new Retext().use(language);

/**
 * Fixtures.
 */

var fixtures = require('./fixtures.js');

/**
 * DOM elements.
 */

var $input = document.getElementsByTagName('textarea')[0];
var $output = document.getElementsByTagName('output')[0];

/**
 * Event handlers
 */

function oninputchange() {
    retext.parse($input.value, function (err, tree) {
        if (err) throw err;

        $output.textContent = tree.data.language;
    });
}

/**
 * Attach event handlers.
 */

$input.addEventListener('input', debounce(oninputchange, 50));

/**
 * Set a random fixture.
 */

$input.value = fixtures[Math.floor(Math.random() * fixtures.length)];

/**
 * Provide initial answer.
 */

oninputchange();


benchmark.js
'use strict';

var Benchmark,
    Retext,
    language;

/*
 * Dependencies.
 */

Benchmark = require('benchmark');
Retext = require('retext');
language = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithLanguage;

retext = new Retext();
retextWithLanguage = new Retext().use(language);

/*
 * Test data: A (big?) section (w/ 10 paragraphs, 50
 * sentences, 1,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

/*
 * Benchmarks
 */

/**
 * Log on start.
 */
function onstart() {
    console.log();
    console.log(this.name);
}

/**
 * Log on cycle.
 *
 * @param {Event} event
 */
function oncycle(event) {
    console.log(String(event.target));
}

var options,
    suite;

options = {
    'async': true,
    'onStart': onstart,
    'onCycle': oncycle
};

suite = new Benchmark.Suite(
    'A paragraph (5 sentences, 100 words)',
    options
);

suite.add('retext w/o retext-language', function () {
    retext.parse(paragraph, function () {});
});

suite.add('retext w/ retext-language', function () {
    retextWithLanguage.parse(paragraph, function () {});
});

suite.run();

suite = new Benchmark.Suite(
    'A section (10 paragraphs, 50 sentences, 1,000 words)',
    options
);

suite.add('retext w/o retext-language', function () {
    retext.parse(section, function () {});
});

suite.add('retext w/ retext-language', function () {
    retextWithLanguage.parse(section, function () {});
});

suite.run();
